// diamonds-typescript/dist/components/images/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.home = exports.close = exports.wall = exports.teleporter = exports.robot = exports.redButton = exports.diamondRed = exports.diamond = exports.botDiamond = exports.botBaseDiamond = exports.botBase = exports.base = exports.baseDiamond = exports.etimoLogo = void 0;
const baseDiamond_svg_1 = __importDefault(require("./baseDiamond.svg"));
exports.baseDiamond = baseDiamond_svg_1.default;
const botBase_svg_1 = __importDefault(require("./botBase.svg"));
exports.botBase = botBase_svg_1.default;
const botBaseDiamond_svg_1 = __importDefault(require("./botBaseDiamond.svg"));
exports.botBaseDiamond = botBaseDiamond_svg_1.default;
const botDiamond_svg_1 = __importDefault(require("./botDiamond.svg"));
exports.botDiamond = botDiamond_svg_1.default;
const close_svg_1 = __importDefault(require("./close.svg"));
exports.close = close_svg_1.default;
const diamond_svg_1 = __importDefault(require("./diamond.svg"));
exports.diamond = diamond_svg_1.default;
const diamondRed_svg_1 = __importDefault(require("./diamondRed.svg"));
exports.diamondRed = diamondRed_svg_1.default;
const etimoLogo_png_1 = __importDefault(require("./etimoLogo.png"));
exports.etimoLogo = etimoLogo_png_1.default;
const home_svg_1 = __importDefault(require("./home.svg"));
Object.defineProperty(exports, "base", { enumerable: true, get: function () { return home_svg_1.default; } });
Object.defineProperty(exports, "home", { enumerable: true, get: function () { return home_svg_1.default; } });
const redButton_svg_1 = __importDefault(require("./redButton.svg"));
exports.redButton = redButton_svg_1.default;
const robot_svg_1 = __importDefault(require("./robot.svg"));
exports.robot = robot_svg_1.default;
const teleporter_svg_1 = __importDefault(require("./teleporter.svg"));
exports.teleporter = teleporter_svg_1.default;
const wall_svg_1 = __importDefault(require("./wall.svg"));
exports.wall = wall_svg_1.default;
