// @digikare/nestjs-prom/dist/prom-core.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var PromCoreModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const prom_constants_1 = require("./prom.constants");
const client = require("prom-client");
const prom_client_1 = require("prom-client");
const prom_utils_1 = require("./common/prom.utils");
let PromCoreModule = PromCoreModule_1 = class PromCoreModule {
    constructor(moduleRef) {
        this.moduleRef = moduleRef;
    }
    static forRoot(options = {}) {
        const { withDefaultsMetrics, registryName, prefix, } = options;
        const promRegistryName = registryName ?
            prom_utils_1.getRegistryName(registryName)
            : prom_constants_1.DEFAULT_PROM_REGISTRY;
        const promRegistryNameProvider = {
            provide: prom_constants_1.PROM_REGISTRY_NAME,
            useValue: promRegistryName,
        };
        const promRegistryOptionsProvider = {
            provide: prom_constants_1.DEFAULT_PROM_OPTIONS,
            useValue: options,
        };
        const registryProvider = {
            provide: promRegistryName,
            useFactory: () => {
                let registry = client.register;
                if (promRegistryName !== prom_constants_1.DEFAULT_PROM_REGISTRY) {
                    registry = new prom_client_1.Registry();
                }
                registry.clear();
                if (options.defaultLabels) {
                    registry.setDefaultLabels(options.defaultLabels);
                }
                if (withDefaultsMetrics !== false) {
                    const defaultMetricsOptions = {
                        register: registry,
                    };
                    if (prefix) {
                        defaultMetricsOptions.prefix = prefix;
                    }
                    prom_client_1.collectDefaultMetrics(defaultMetricsOptions);
                }
                return registry;
            },
        };
        return {
            module: PromCoreModule_1,
            providers: [
                promRegistryNameProvider,
                promRegistryOptionsProvider,
                registryProvider,
            ],
            exports: [
                registryProvider,
            ],
        };
    }
    onModuleDestroy() {
    }
};
PromCoreModule = PromCoreModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({}),
    __metadata("design:paramtypes", [core_1.ModuleRef])
], PromCoreModule);
exports.PromCoreModule = PromCoreModule;
